aws-java-sdk-core/src/main/java/com/amazonaws/auth/profile/internal/securitytoken/STSProfileCredentialsServiceProvider.java

```diff
@@ -0,0 +1,64 @@
+package com.amazonaws.auth.profile.internal.securitytoken;
+
+import com.amazonaws.AmazonClientException;
+import com.amazonaws.auth.AWSCredentials;
+import com.amazonaws.auth.AWSCredentialsProvider;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+
+/**
+ *
+ */
+public class STSProfileCredentialsServiceProvider implements AWSCredentialsProvider {
+    private static final String CLASS_NAME = "com.amazonaws.services.securitytoken.internal.STSProfileCredentialsService";
+    private static Constructor<ProfileCredentialsService> STS_SERVICE_CONSTRUCTOR;
+
+    private final RoleInfo roleInfo;
+    private AWSCredentialsProvider profileCredentialsProvider;
+
+    public STSProfileCredentialsServiceProvider(RoleInfo roleInfo) {
+        this.roleInfo = roleInfo;
+    }
+
+    private AWSCredentialsProvider getProfileCredentialsProvider() {
+        if (this.profileCredentialsProvider == null) {
+            this.profileCredentialsProvider = getProfileCredentialService().getAssumeRoleCredentialsProvider(roleInfo);
+        }
+
+        return this.profileCredentialsProvider;
+    }
+
+    private static ProfileCredentialsService getProfileCredentialService() {
+        if (STS_SERVICE_CONSTRUCTOR == null) {
```

Comment:
Sorry this was more my motiviation in making the Loader a singleton. I'd rather not have to use reflection more than once. We can lazily instantiate the ProfileCredentialsService and cache it in the singleton reference.